packages/ai-jsx/src/core/opentelemetry.ts

```diff
@@ -0,0 +1,129 @@
+import opentelemetry, { SpanContext, SpanStatusCode, context, createContextKey } from '@opentelemetry/api';
+import { PartiallyRendered, StreamRenderer } from './render.js';
+import { Element, isElement } from './node.js';
+import { memoizedIdSymbol } from './memoize.js';
+import { debug } from './debug.js';
+
+function bindAsyncGeneratorToActiveContext<T = unknown, TReturn = any, TNext = unknown>(
+  generator: AsyncGenerator<T, TReturn, TNext>
+): AsyncGenerator<T, TReturn, TNext> {
+  const activeContext = context.active();
+  const result = {
+    next: context.bind(activeContext, generator.next.bind(generator)),
+    return: context.bind(activeContext, generator.return.bind(generator)),
+    throw: context.bind(activeContext, generator.throw.bind(generator)),
+
+    [Symbol.asyncIterator]() {
+      return result;
+    },
+  };
+
+  return result;
+}
+
+// Tracks all memoized subtrees that are being actively traced.
+const memoizedIdsToActiveSpanContexts = new Map<number, SpanContext>();
+
+interface MemoizedIdAndOwner {
+  id: number;
+  isOwner: boolean;
+}
+
+// Indicates whether the current async context is responsible for tracing the memoized subtree.
+const activeMemoizedIdStorage = createContextKey('The active MemoizedIdAndOwner');
+
+function spanAttributesForElement(element: Element<any>): opentelemetry.Attributes {
+  return { 'ai.jsx.tag': element.tag.name, 'ai.jsx.tree': debug(element, true) };
+}
+
+export function openTelemetryStreamRenderer(streamRenderer: StreamRenderer): StreamRenderer {
```

Comment:
Will this show up in the generated API docs? I slightly suspect that every `export` needs either `@hidden` or a JSDoc comment.

Reply:
It won't because it's not an `entryPoint` in the typedoc section of the tsconfig.json. (It's also not externally exported at all.)